src/ui/page/share/dataset/steps/metadataFilling/metadataFilling.tsx

```diff
@@ -0,0 +1,346 @@
+/* eslint-disable max-lines-per-function */
+import { useEffect, useMemo, useCallback, type FC } from 'react'
+import { useTranslation } from 'react-i18next'
+import * as O from 'fp-ts/Option'
+import * as IOE from 'fp-ts/IOEither'
+import * as S from 'fp-ts/string'
+import { flow, pipe } from 'fp-ts/lib/function'
+import type {
+  FormItem,
+  FormItemType,
+  FormItemValue,
+  initFormPayload
+} from '@/ui/store/slice/shareData/shareData.slice'
+import { useAppStore } from '@/ui/store'
+import type { ResourceError } from '@/shared/error'
+import { ShowFileError } from '@/shared/error'
+import { Field } from '@/ui/component/field/field'
+import type { NotificationType } from '@/ui/component/notification/notification'
+import { useDispatchNotification } from '@/ui/hook/useDispatchNotification'
+import './metadataFilling.scss'
+
+type DatasetFormItem = {
+  id: string
+  type: FormItemType
+  label: string
+  value: FormItemValue
+  render: () => JSX.Element
+  required?: boolean
+  style?: {
+    side?: 'left' | 'right'
+    inRow?: boolean
+  }
+}
+
+type DatasetForm = DatasetFormItem[]
+
+type NotificationData = {
+  title: string
+  type: NotificationType
+}
+
+const formErrorData = (error: ResourceError): NotificationData => {
+  switch (error._tag) {
+    case 'resource-not-found':
+    case 'resource-already-exists':
+      return {
+        title: 'notification:error.problem',
+        type: 'error'
+      }
+  }
+}
+
+export const MetadataFilling: FC = () => {
+  const { t } = useTranslation('share')
+  const dispatchNotification = useDispatchNotification()
+  const { initForm, setFormItemValue, formItemById, isFormInitialized } = useAppStore(state => ({
+    initForm: state.shareData.initForm,
+    setFormItemValue: state.shareData.setFormItemValue,
+    formItemById: state.shareData.formItemById,
+    isFormInitialized: state.shareData.isFormInitialized
+  }))
+
+  const formSides = ['left', 'right']
+
+  const handleFieldValueChange = useCallback(
+    (id: string) => (value: string) => {
+      setFormItemValue({ id, value })()
+    },
+    [setFormItemValue]
+  )
+
+  const formItemFieldValue = useCallback(
+    (id: string): string =>
+      flow(
+        formItemById(id),
+        O.map(({ value }: FormItem) => value),
+        O.chain(O.fromPredicate(S.isString)),
+        O.getOrElse(() => '')
+      )(),
+    [formItemById]
+  )
+
+  const datasetForm: DatasetForm = useMemo((): DatasetForm => {
+    const id1 = 'input-field-1'
```

Comment:
It's correct, but it would have been better to place the declarations as close as possible to their use, in each element of the array (since they are independent each other), but this would make the syntax a bit cumbersome.

Reply:
@ccamel 
If you don't mind, I'll suggest to keep it like that to prevent from a heavy syntax in the `useMemo` function.
Do you agree?

Reply:
Of course.